### ISSUE TITLE
Crash during plan with v2.9.0

### ISSUE THREAD

**Issue by pdecat** [terraform-providers/terraform-provider-google#3884]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version

```
# terraform version
Terraform v0.12.2
+ provider.google v2.9.0
+ provider.google-beta v2.9.0
```

### Affected Resource(s)

* N/A

### Terraform Configuration Files

```tf
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://www.hashicorp.com/security
# If reproducing the bug involves modifying the config file (e.g., apply a config,
# change a value, apply the config again, see the bug) then please include both the
# version of the config before the change, and the version of the config after the change.
```

### Panic Output

https://gist.github.com/pdecat/e74062aa03e83a335291eb767ab48039

### Steps to Reproduce

1. `terraform init && terraform plan`

### Important Factoids

Just upgrading from v2.8.0 to v2.9.0

### References

* N/A

**Comment by pdecat** [comment]
Downgrading to 2.8.0 works.

**Comment by danawillow** [comment]
@pdecat, do you have the tf config for the resource that this is failing on? Would be helpful to make sure we have a test that covers that case

**Comment by pdecat** [comment]
Gist deleted, needs redaction.

**Comment by pdecat** [comment]
To my understanding, it is crashing at https://github.com/terraform-providers/terraform-provider-google/blame/v2.9.0/google/resource_storage_bucket.go#L527
so either `res.IamConfiguration` is nil or `res.IamConfiguration.BucketPolicyOnly` is.

**Comment by danawillow** [comment]
Yes, I can see that. However, we have tests that set bucket_policy_only and tests that don't, and none of them are failing, so I don't know how to reproduce the crash.

**Comment by pdecat** [comment]
Here's the bucket configuration:

```
resource "google_storage_bucket" "terraform_root_state_bucket" {
  name     = "gcs_terraform_root-state_${var.general["region"]}_${var.general["project"]}"
  location = "${upper(var.general["region"])}"

  versioning {
    enabled = "true"
  }
}
```

**Comment by danawillow** [comment]
Thanks! I'll see if I can repro and get a fix out today.

**Comment by pdecat** [comment]
Looking at https://github.com/terraform-providers/terraform-provider-google/pull/3783/files#diff-b93d962cfcd3c90fba7b47c0240bd1fcR798
it is most certain that it is `res.IamConfiguration.BucketPolicyOnly` that is nil.

**Comment by pdecat** [comment]
Here's what the API returns:

```
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4: {
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "kind": "storage#bucket",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "id": "gcs_terraform_root-state_europe-west1_myproject",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "selfLink": "https://www.googleapis.com/storage/v1/b/gcs_terraform_root-state_europe-west1_myproject",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "projectNumber": "36727164259",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "name": "gcs_terraform_root-state_europe-west1_myproject",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "timeCreated": "2016-09-30T16:10:57.955Z",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "updated": "2018-01-16T16:31:45.868Z",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "metageneration": "4",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "location": "EUROPE-WEST1",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "versioning": {
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:   "enabled": true
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  },
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "storageClass": "STANDARD",
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4:  "etag": "CAQ="
2019-06-20T00:09:55.454+0200 [DEBUG] plugin.terraform-provider-google_v2.9.0_x4: }
```

**Comment by danawillow** [comment]
Weird! When I send a request that doesn't set IamConfiguration, it still gets returned. Regardless, it's not hard to add nil checks.

Proof (removing some headers for clarity):
```
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: ---[ REQUEST ]---------------------------------------
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: POST /storage/v1/b?alt=json&prettyPrint=false&project=graphite-test-danahoffman-tf HTTP/1.1
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: Host: www.googleapis.com
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: User-Agent: google-api-go-client/0.5 Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: Content-Length: 115
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: Content-Type: application/json
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: Accept-Encoding: gzip
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: 
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: {
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  "lifecycle": {
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:   "rule": []
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  },
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  "location": "US",
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  "name": "is-3884",
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  "storageClass": "STANDARD",
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  "versioning": {
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:   "enabled": true
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google:  }
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: }
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: 
2019-06-19T15:11:37.335-0700 [DEBUG] plugin.terraform-provider-google: -----------------------------------------------------
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google: 2019/06/19 15:11:38 [DEBUG] Google API Response Details:
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google: ---[ RESPONSE ]--------------------------------------
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google: HTTP/2.0 200 OK
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google: 
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google: {
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "kind": "storage#bucket",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "id": "is-3884",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "selfLink": "https://www.googleapis.com/storage/v1/b/is-3884",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "projectNumber": "515892857929",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "name": "is-3884",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "timeCreated": "2019-06-19T22:11:38.095Z",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "updated": "2019-06-19T22:11:38.095Z",
2019-06-19T15:11:38.352-0700 [DEBUG] plugin.terraform-provider-google:  "metageneration": "1",
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  "iamConfiguration": {
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:   "bucketPolicyOnly": {
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:    "enabled": false
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:   }
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  },
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  "location": "US",
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  "versioning": {
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:   "enabled": true
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  },
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  "storageClass": "STANDARD",
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google:  "etag": "CAE="
2019-06-19T15:11:38.353-0700 [DEBUG] plugin.terraform-provider-google: }
```

**Comment by pdecat** [comment]
Maybe it does that only for new buckets, ours are somewhat old.

**Comment by pdecat** [comment]
Can confirm the issue is resolved with v2.9.1.

Thanks Dana!
